backend/selectgen.ml

```diff
@@ -28,18 +28,52 @@ type trap_stack_info =
   | Unreachable
   | Reachable of trap_stack
 
-type region_stack = Reg.t array list
+module Region_stack : sig
+  (* A nested set of regions, with the innermost at the head *)
+  type t = Reg.t array list
```

Comment:
I think you can make this abstract and just add `extend` (i.e. cons), `outermost` and `innermost` functions.

Reply:
The other operation you are using on these is `==`, which I'm a bit worried about. It's fine to use that for optimisation, but you're using it in `assertions` and OCaml's "definition" pretty clearly allows it to compile this file into something which would fail those assertions because lists are immutable. Could you perhaps add a`same` function that uses `==` on the array in the head of the list instead? You could use that one in the code for `Cregion` too and then not add the `innermost` function I proposed above.